### DISCUSSION TITLE
nestJs=true + outputServices=grpc-js

### DISCUSSION THREAD

**Discussion by lautarodragan** [stephenh/ts-proto#784] [Q&A] [answered]
Hey there. Thanks for this awesome library!

I'm currently in need of generating the `.ts` from my `.proto` with both of these options enabled.

I saw in the code that this is not possible. The `outputServices` are ignored if `nestJs` is truthy:

https://github.com/stephenh/ts-proto/blob/3700d48e8ebd2ac4ed6cb0e3a7bf8fcb3e5f993d/src/main.ts#L249-L268

I created a PR to update the documentation regarding this: https://github.com/stephenh/ts-proto/pull/783

But I do have a question:

Since I need the output of both options, I'm currently running `protoc` twice, like this:

```

protoc \
  --experimental_allow_proto3_optional \
  --plugin=node_modules/ts-proto/protoc-gen-ts_proto \
  --ts_proto_opt=env=node \
  --ts_proto_opt=nestJs=true \
  --ts_proto_opt=stringEnums=true \
  --ts_proto_opt=addGrpcMetadata=true \
  --ts_proto_opt=addNestjsRestParameter=true \
  --ts_proto_opt=esModuleInterop=true \
  --ts_proto_opt=forceLong=long \
  --ts_proto_opt=unrecognizedEnum=false \
  --ts_proto_opt=outputSchema=true \
  --ts_proto_out="$PROTO_PATH" \
  "$PROTO_PATH"/*.proto \
  -I"$PROTO_PATH"/.

mkdir -p "$PROTO_PATH/grpc"

protoc \
  --experimental_allow_proto3_optional \
  --plugin=node_modules/ts-proto/protoc-gen-ts_proto \
  --ts_proto_opt=outputServices=grpc-js \
  --ts_proto_opt=forceLong=string \
  --ts_proto_out="$PROTO_PATH/grpc" \
  "$PROTO_PATH"/*.proto \
  -I"$PROTO_PATH"/.

```

Basically, creating one set of outputs in `./` and another in `./grpc/`.

As a cheap quality-of-life improvement, to avoid importing from two separate files, I'm adding an `export & from './grpc/${fileName}` to every generated `.ts` file from the first pass.

So, for example, instead of doing this:

```
import { Hero } from './hero';
import { HeroServiceService } from './grpc/hero';
```

I can do this:

```
// in /hero.ts

export * from './grpc/hero';

// in my file

import { Hero, HeroServiceService } from './hero';
```

With the case I tried, this worked fine. But I just got lucky there are declaration conflicts between one file and the other, otherwise the `export * from` would blow up.

So, my question is: is this likely to blow up soon, or is it normal for there not to be conflicts between files?

**Comment by stephenh** [comment, marked as answer]
Fwiw @lautarodragan I'm pretty sure that "if `nestJs=true` ignore the services" was not necessarily on purpose, but just a happenstance that, when the `nestJs` code was originally written, the author didn't want any of the existing output/services, and the `if/else` was the easiest way to do it.

If you'd like to tinker with the code, and be able to output both nestJs=true + outputServices=grpc-js that seems fine to me? I guess?

Out of curiosity, is this b/c you're like using NestJS on the client-side but not-NestJS on the server-side? Or the other way around?

**Reply by lautarodragan** [reply]
Thank you @stephenh for providing insight on this!

It's actually because we set up a "mock" server for our tests that has hard-coded responses, and this mock server doesn't use NestJS at all.

> If you'd like to tinker with the code, and be able to output both nestJs=true + outputServices=grpc-js that seems fine to me? I guess?

Oki, I'll see if I can tackle this.
